src/components/map/Graph.js

```diff
@@ -51,106 +54,103 @@ const Title = styled.div`
   padding: 5px;
 `;
 
-@inject(app("Filters", "UI"))
-@observer
-class Graph extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {
-      highlight: {x: 0, y: 0},
-      avgSpeed: {x: 0, y: 0},
-    };
-  }
-
-  secondsToTime() {
-    let diff = this.state.highlight.y;
-    const departureDiffSign = diff < 0 ? "-" : "";
-    diff = Math.abs(diff);
-    let minutes = Math.floor(diff / 60);
-    let seconds = diff - minutes * 60;
-    minutes = doubleDigit(minutes);
-    seconds = doubleDigit(seconds);
-    return departureDiffSign + minutes + ":" + seconds;
-  }
-
-  coloredBackgroundSlotColor() {
-    return this.state.highlight.departureColor
-      ? this.state.highlight.departureColor
-      : "var(--light-grey)";
-  }
-
-  render() {
-    const {diffs, speedAverages, graphExpanded, UI, width} = this.props;
-
-    const highlight = this.state.highlight;
-    const avgSpeed = this.state.avgSpeed;
-    const mapDiffs = diffs.map(function(o) {
-      return o.y;
-    });
+const decorate = flow(
+  observer,
+  inject("Filters", "UI")
+);
+
+const Graph = decorate((props) => {
+  const {departures, events, graphExpanded, UI, Filters, width} = props;
+
+  const diffs = useMemo(() => getJourneyStopDiffs(departures), [departures]);
+  const speedAverages = useMemo(() => getJourneyAverageSpeeds(events), [events]);
+
+  const [highlight, setHighlight] = useState({x: 0, y: 0});
+  const [avgSpeed, setAvgSpeed] = useState({x: 0, y: 0});
+
+  const mapDiffs = useMemo(
+    () =>
+      diffs.map((o) => {
+        return o.y;
+      }),
+    [diffs]
+  );
+
+  const {max, min} = useMemo(() => {
     const max = Math.max.apply(Math, mapDiffs);
     const min = Math.min.apply(Math, mapDiffs);
+    return {max, min};
+  }, [mapDiffs]);
 
-    return (
-      <div>
-        {graphExpanded && width && (
-          <XYPlot
-            height={200}
-            width={width}
-            onClick={(d) => {
-              this.props.Filters.setStop(highlight.stopId);
+  const coloredBackgroundSlotColor = useMemo(() => {
+    return highlight.departureColor ? highlight.departureColor : "var(--light-grey)";
+  }, [highlight]);
+
+  const time = useMemo(() => secondsToTime(highlight.y), [highlight]);
+
+  useEffect(() => {
+    if (diffs.length !== 0 && highlight.x === 0 && highlight.y === 0) {
+      setHighlight(diffs[0]);
+    }
+
+    if (speedAverages.length !== 0 && avgSpeed.x === 0 && avgSpeed.y === 0) {
+      const data = speedAverages[0];
+      const graphPoint = {x: data.x, y: data.y};
+      setAvgSpeed(graphPoint);
+    }
+  }, [diffs, speedAverages]);
```

Comment:
The useEffect hook is the main reason I refacatored this component. It makes use cases like setting initial values very easy and performant.